clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java

```diff
@@ -37,7 +36,7 @@
  *
  * <em>Note</em>: this class is not thread-safe and is intended to only be used from a single thread.
  */
-public class FetchBuffer implements Closeable {
+public class FetchBuffer implements AutoCloseable {
```

Comment:
Yes, the process is a bit convoluted...

To perform the process of moving the fetched records from the background thread to the application thread and then on to the user, `PrototypeAsyncConsumer` has these three instance variables:

1. `fetchResults`
2. `fetchBuffer`
3. `fetchCollector`

All three of those objects are created in the application thread when the `PrototypeAsyncConsumer` is created. `fetchBuffer` and `fetchCollector` are only ever referenced by the application thread; `fetchResults`, however, is used by **both** threads.

`fetchResults` is referenced in the background thread when it is used in the `FetchEvent` callback in the `sendFetches()` method:

```java
private void sendFetches() {
    FetchEvent event = new FetchEvent();
    applicationEventHandler.add(event);

    event.future().whenComplete((completedFetches, error) -> {
        fetchResults.addAll(completedFetches);
    });
}
```

Since the `whenComplete()` method is executed when the background thread "completes" the `Future`, `fetchResults` is thus modified on the background thread.

The rest of the process should occur on the application thread.

During calls to `poll()` on the application thread, data from `fetchResults` is moved to `fetchBuffer` in `pollForFetches()`:

```java
while (pollTimer.notExpired()) {
    CompletedFetch completedFetch = fetchResults.poll(pollTimer.remainingMs(), TimeUnit.MILLISECONDS);

    if (completedFetch != null)
        fetchBuffer.add(completedFetch);

    pollTimer.update();
}
```

The data in `fetchBuffer` is later extracted in `fetchCollector` during the `poll()` process, but this again is on the application thread.

This roundabout way of getting the data is specifically done so that we don't write to the `FetchBuffer` inadvertently from the background thread. Hence these JavaDoc comment for `fetchResults`:

```java
/**
 * A thread-safe {@link BlockingQueue queue} for the results that are populated in the background thread
 * when the fetch results are available. Because the {@link #fetchBuffer fetch buffer} is not thread-safe, we
 * need to separate the results collection that we provide to the background thread from the collection that
 * we read from on the application thread.
 */
private final BlockingQueue<CompletedFetch> fetchResults = new LinkedBlockingQueue<>();
```

This is a rough idea of what happens on the background thread:

>`FetchRequestManager` -> `fetchResults`

Then later in the application thread during `poll()`:

> `fetchResults` -> `fetchBuffer` -> `fetchCollector`

Let me know if that makes sense or if there is still a gap that I'm not seeing. I can write the above up (with any changes you'd like) in code comments.

Reply:
@kirktrue : Thanks for the explanation. Two followup questions.
1. The background thread has the following path` FetchRequestManager.poll  -> handleFetchResponse -> fetchBuffer.add(completedFetch)`. So, it seems that the background thread also writes the fetched data to `fetchBuffer`.
2. This is related to my other [comment](https://github.com/apache/kafka/pull/14406#discussion_r1342984775). The background thread has the following path `FetchRequestManager.poll  -> AbstractFetch.prepareFetchRequests -> AbstractFetch.fetchablePartitions -> reads fetchBuffer.bufferedPartitions()`. Since fetchBuffer is written by the application thread, how do we coordinate the synchronization btw the two threads?

Reply:
For point #1, the background thread has a _separate_ fetch buffer. It doesn't write to the same object. The application thread `FetchBuffer` is created in `PrototypeAsyncConsumer` and the background thread `FetchBuffer` is created in `AbstractFetch`.

For point #2, I think your [other comment](https://github.com/apache/kafka/pull/14406#discussion_r1342984775) is correct. I'm not yet sure how to maintain those two fetch buffers separately without running into that race condition.

Reply:
I just pushed a proposed fix for this, which is basically to make the `FetchBuffer` thread safe. Now there is only one fetch buffer for the consumer and is accessed by both threads.

Reply:
@junrao this is one of the more serious and subtle issues, so I want to know you're OK with the new approach before I resolve this conversation in the PR. Thanks!

Reply:
@kirktrue : Yes, the fix LGTM.

Reply:
Great 😄 Thanks!